File: DSA/done/1046 - lastStoneWeight.cpp
// https://leetcode.com/problems/last-stone-weight/

#include <queue>
#include <vector>

using namespace std;

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> maxHeap;
        for (int s : stones) {
            maxHeap.push(s);
        }

        while (maxHeap.size() > 1) {
            int first = maxHeap.top();
            maxHeap.pop();
            int second = maxHeap.top();
            maxHeap.pop();
            if (second < first) {
                maxHeap.push(first - second);
            }
        }

        return maxHeap.empty() ? 0 : maxHeap.top();
    }
};

File: DSA/done/1046 - lastStoneWeight.py
# https://leetcode.com/problems/last-stone-weight/

import heapq
from typing import List

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        maxHeap = [-s for s in stones]
        heapq.heapify(maxHeap)
        while len(maxHeap) > 1:
            s1, s2 = -heapq.heappop(maxHeap), -heapq.heappop(maxHeap)
            if s1 > s2:
                heapq.heappush(maxHeap, -(s1 - s2))
        return -maxHeap[0] if maxHeap else 0